ingest/src/main/scala/hydra.ingest/http/IngestionEndpoint.scala

```diff
@@ -121,92 +109,123 @@ class IngestionEndpoint[F[_]: Futurable](
-              case Failure(other) =>
-                val responseCode = StatusCodes.ServiceUnavailable
-                val errorMsg =
-                  s"Exception: $other; ${hydraRequest.correlationId}: Ack:${hydraRequest.ackStrategy}; Validation: ${hydraRequest.validationStrategy};" +
-                    s" Metadata:${hydraRequest.metadata}; Payload: ${hydraRequest.payload} Ingestors: Alt-Ingest-Flow"
-                log.error(s"Ingestion failed for request $errorMsg")
-                addPromMetric(topic, StatusCodes.ServiceUnavailable.toString())
-                complete(responseCode, IngestionReport(hydraRequest.correlationId, Map("kafka_ingestor" -> IngestorError(other)), responseCode.intValue))
-            }
-          } else {
-            imperativelyComplete { ctx =>
-              requestHandler ! InitiateHttpRequest(
-                hydraRequest,
-                ingestTimeout,
-                ctx
-              )
-            }
+          handleExceptions(exceptionHandler(topic)) {
+            if (alternateIngestFlowEnabled && useAlternateIngestFlow(hydraRequest)) {
+              sendAltFlow(hydraRequest,topic)
+            } else {
+              imperativelyComplete { ctx =>
+                requestHandler ! InitiateHttpRequest(
+                  hydraRequest,
+                  ingestTimeout,
+                  ctx
+                )
+              }
           }
         }
       }
     }
   }
 
-  private def exceptionHandler = ExceptionHandler {
-    case e: IllegalArgumentException =>
-      if (applicationConfig
+  private def exceptionHandler(topic: String) = ExceptionHandler {
+      case e: IllegalArgumentException =>
+        extractExecutionContext { implicit ec =>
+          if (applicationConfig
             .getBooleanOpt("hydra.ingest.shouldLog400s")
             .getOrElse(false)) {
-        log.error("Ingestion 400 ERROR: " + e.getMessage)
-      }
-      complete(400, GenericError(400, e.getMessage))
+            log.error("Ingestion 400 ERROR: " + e.getMessage)
+          }
+          addPromHttpMetric(topic, StatusCodes.BadRequest.toString,"ingestionEndpoint")
+          complete(400, GenericError(400, e.getMessage))
+        }
+
+      case e: Throwable =>
+        extractExecutionContext { implicit ec =>
+          addPromHttpMetric(topic, StatusCodes.InternalServerError.toString,"ingestionEndpoint")
+          complete(500, GenericError(500, e.getMessage))
+        }
+
+      case e: Exception =>
+        extractExecutionContext { implicit ec =>
+          addPromHttpMetric(topic, StatusCodes.InternalServerError.toString,"ingestionEndpoint")
+          complete(500, GenericError(500, e.getMessage))
+        }
+
+      case e: Error =>
+        extractExecutionContext { implicit ec =>
+          addPromHttpMetric(topic, StatusCodes.InternalServerError.toString,"ingestionEndpoint")
+          complete(500, GenericError(500, e.getMessage))
+        }
```

Comment:
I am thinking that this is a redundant way of writing this handler. You could probably just match on `case e =>` and get all of those types in one shot rather than separately.